File: java/sort/array/ShellSort.java
package sort.array;

public class ShellSort {
    public void sort(int[] nums) {
        int gap = 1;
        while (gap < nums.length / 3) {
            gap = gap * 3 + 1;
        }

        while (gap > 0) {
            for (int i = gap; i < nums.length; ++i) {
                int pivot = nums[i];
                int index = i - gap;
                while (index >= 0 && nums[index] > pivot) {
                    nums[index + gap] = nums[index];
                    index -= gap;
                }
                nums[index + gap] = pivot;
            }
            gap /= 3;
        }
    }
}

File: py/sort/array/ShellSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class ShellSort:
    def sort(self, nums: List[int]) -> None:
        gap = 1
        while gap < len(nums) // 3:
            gap = gap * 3 + 1

        while gap > 0:
            for i, value in enumerate(nums[gap : ]):
                pivot, index = value, i
                while index >= 0 and nums[index] > pivot:
                    nums[index + gap] = nums[index]
                    index -= gap
                nums[index + gap] = pivot
            gap //= 3
